compiler/src/dotty/tools/dotc/transform/ShortcutImplicits.scala

```diff
@@ -0,0 +1,162 @@
+     *  A fresh map is created for each unit.
+     */
+    private val directMeth = new mutable.HashMap[Symbol, Symbol]
+
+    /** Should `sym` get a ..$direct companion?
+     *  This is the case if (1) `sym` is a method with an implicit function type as final result type.
+     *  However if `specializeMonoTargets` is true, we exclude symbols that are known
+     *  to be only targets of monomorphic calls because they are effectively
+     *  final and don't override anything.
+     */
+    private def shouldBeSpecialized(sym: Symbol)(implicit ctx: Context) =
+      sym.is(Method, butNot = Accessor) &&
+      defn.isImplicitFunctionType(sym.info.finalResultType) &&
+      (specializeMonoTargets || !sym.isEffectivelyFinal || sym.allOverriddenSymbols.nonEmpty)
+
+    /** @pre    The type's final result type is an implicit function type `implicit Ts => R`.
+     *  @return The type of the `apply` member of `implicit Ts => R`.
+     */
+    private def directInfo(info: Type)(implicit ctx: Context): Type = info match {
+      case info: PolyType   => info.derivedPolyType(resType = directInfo(info.resultType))
+      case info: MethodType => info.derivedMethodType(resType = directInfo(info.resultType))
+      case info: ExprType   => directInfo(info.resultType)
+      case info             => info.member(nme.apply).info
+    }
+
+    /** A new `m$direct` method to accompany the given method `m` */
+    private def newDirectMethod(sym: Symbol)(implicit ctx: Context): Symbol =
+      sym.copy(
+        name = sym.name.directName,
+        flags = sym.flags | Synthetic,
+        info = directInfo(sym.info))
+
+    /** The direct method `m$direct` that accompanies the given method `m`.
+     *  Create one if it does not exist already.
+     */
+    private def directMethod(sym: Symbol)(implicit ctx: Context): Symbol =
+      if (sym.owner.isClass) {
+        val direct = sym.owner.info.member(sym.name.directName)
+          .suchThat(_.info matches directInfo(sym.info)).symbol
+        if (direct.maybeOwner == sym.owner) direct
+        else newDirectMethod(sym).enteredAfter(thisTransform)
+      }
+      else directMeth.getOrElseUpdate(sym, newDirectMethod(sym))
+
+
+    /** Transform `qual.apply` occurrences according to rewrite rule (2) above */
+    override def transformSelect(tree: Select)(implicit ctx: Context, info: TransformerInfo) =
+      if (tree.name == nme.apply &&
+          defn.isImplicitFunctionType(tree.qualifier.tpe.widen) &&
+          shouldBeSpecialized(tree.qualifier.symbol)) {
+        def directQual(tree: Tree): Tree = tree match {
+          case Apply(fn, args)     => cpy.Apply(tree)(directQual(fn), args)
+          case TypeApply(fn, args) => cpy.TypeApply(tree)(directQual(fn), args)
+          case Block(stats, expr)  => cpy.Block(tree)(stats, directQual(expr))
+          case tree: RefTree =>
+            cpy.Ref(tree)(tree.name.directName)
+              .withType(directMethod(tree.symbol).termRef)
```

Comment:
I'm guessing that this transform is (correctly) skipped for:

```
scala> def foo(a: Any): Transactional[Int] = 42
def foo(a: Any): Transactional[Int]
scala> (if ("".isEmpty) foo("") else foo("")).apply("")
```

Because `tree.qualifier.symbol` would be `NoSymbol`.

But just wanted to check that I'm reading this correctly.